### ISSUE TITLE
Per job ACLs, vault policies ACLs, host_volume use ACLs

### ISSUE THREAD

**Issue by urusha** [hashicorp/nomad#8865]
It would be great if we could specify per job ACL policies. So that particular token could only `read-fs`/`alloc-exec`/etc particular job (jobs with prefix).
This is the main feature we'd like to be added, but there are some other issues which should be addressed to get full control of permissions on per job basis. Some of them are actually security issues. If at least part of them would get implemented that would be really nice improvement!
1. Vault policies ACLs. A token should be allowed to only use particular vault-policies and this should affect namespace policies: `submit-job`, `dispatch-job`, `alloc-exec`, `read-fs`(?). So that token without sufficient vault permissions would not be able to execute commands inside running jobs (which equals to having access to /secrets or ENV variables). This is a security issue, if token has `alloc-exec` permission (to any job with the vault policy), it has access to this vault policy resources.
2. Host volumes ACLs. Much like vault policies, if the token has `alloc-exec`/`read-fs`/etc permission on the job, it must also have access to all volumes mounted in this job, to actually use `alloc-exec` or `read-fs`.
3. Both types of ACLs might have `use` policy to preserve current behaviour which allows `alloc-exec` without vault/host_volume policies. Like this
```
vault "vaultgroup-*" {
  policy = "use"
}
host_volume "volgroup-*" {
  policy = "use"
}
```

With 1,2 implemented, `alloc-exec` permission and no `submit-job`,`dispatch-job` permission for a token, that token would only be able to execute command inside running job with allowed vault-policies and host_volumes. Dummy vault-policies/host_volumes might be used to variate token's permissions, and this actually would be per job ACLs. And this would be very convenient to allow developers to debug particular jobs in production.
Thanks for attention.

**Comment by tgross** [comment]
Hi @urusha!

> 1. Vault policies ACLs. A token should be allowed to only use particular vault-policies 

Generally the way this is handled is by requiring the job submitter to provide a Vault token. See the Nomad config docs for [`allow_unauthenticated`](https://www.nomadproject.io/docs/configuration/vault#allow_unauthenticated). The Vault token can have whatever policies you need to restrict.

> So that token without sufficient vault permissions would not be able to execute commands inside running jobs (which equals to having access to /secrets or ENV variables). This is a security issue, if token has allow-exec permission (to any job with the vault policy), it has access to this vault policy resources.
> 2. Host volumes ACLs. Much like vault policies, if the token has allow-exec/read-fs/etc permission on the job, it must also have access to all volumes mounted in this job, to actually use allow-exec or read-fs.

I'm not sure I understand the concern here. The job can contain arbitrary code, so the `allow-exec` and `read-fs` permissions doesn't provide any privilege escalation over what the job itself can do. The job could exfiltrate that same secret data without manual intervention.

**Comment by urusha** [comment]
> Hi @urusha!

Hello, @tgross,
thanks for the answer!
> 
> > 1. Vault policies ACLs. A token should be allowed to only use particular vault-policies
> 
> Generally the way this is handled is by requiring the job submitter to provide a Vault token. See the Nomad config docs for [`allow_unauthenticated`](https://www.nomadproject.io/docs/configuration/vault#allow_unauthenticated). The Vault token can have whatever policies you need to restrict.

Didn't know about this feature. Just tested. First - as I understand, current nomad UI doesn't provide support for specifiyng vault token. Second, in our case we don't need to partly restrict submitting of jobs, but rather restrict executing and reading fs of already running jobs. And this leads to third: this setting allows to run `nomad alloc exec` to any already running container - and this is the main problem. Even if my token doesn't allow submitting job with particular vault policy, but I'm granted `alloc-exec` in the namespace - I can get `secrets` and ENV of any job, so I can get access to any running vault-policy (or at least results of templates formed using any vault-policy in the namespace). And this is the security issue, the only current workaround - restrict `alloc-exec` and `read-fs` in the namespace for all untrusted tokens. Please, correct me, if I'm wrong.

Also, specifying allowed vault policy in the nomad policy (like suggested in the issue) looks more convenient (starting with providing vault token in the nomad UI). So that single nomad token describes all permissions (both nomad and vault).

Pretty the same for `host_volumes`. Token might not be able to submit job with particular `host_volume`, but is able to `read-fs`,`alloc-exec` (write), and this affects any job.

> 
> > So that token without sufficient vault permissions would not be able to execute commands inside running jobs (which equals to having access to /secrets or ENV variables). This is a security issue, if token has allow-exec permission (to any job with the vault policy), it has access to this vault policy resources.
> > 2. Host volumes ACLs. Much like vault policies, if the token has allow-exec/read-fs/etc permission on the job, it must also have access to all volumes mounted in this job, to actually use allow-exec or read-fs.
> 
> I'm not sure I understand the concern here. The job can contain arbitrary code, so the `allow-exec` and `read-fs` permissions doesn't provide any privilege escalation over what the job itself can do. The job could exfiltrate that same secret data without manual intervention.

Please, see above.

**Comment by tgross** [comment]
> current nomad UI doesn't provide support for specifiyng vault token.

Ah, that's probably the case. Would you be willing to open a new issue for that specific item?

> Even if my token doesn't allow submitting job with particular vault policy, but I'm granted alloc-exec in the namespace - I can get secrets and ENV of any job, so I can get access to any running vault-policy (or at least results of templates formed using any vault-policy in the namespace)

Right, you can get access to any job within your namespace. The way to restrict that is to use the Namespaces feature (combined with the multi-Vault Namespaces support we shipped in [0.12.2](https://github.com/hashicorp/nomad/blob/master/CHANGELOG.md#0122-august-12-2020). If we were to provide more fine-grained access control, I'd expect that would also land in the Enterprise version of Nomad.

**Comment by tgross** [comment]
Starting in [Nomad 1.7.0-beta.1](https://github.com/hashicorp/nomad/releases/tag/v1.7.0-beta.1) we've deprecated the use of Vault tokens at the time of job submission. Nomad will use workload identities to sign into Vault for purposes of getting Vault tokens for those workloads. Using Vault tokens during job submission will be removed entirely in Nomad 1.9.
